File: cann/cpp/src/kzzk_cv.cpp
#include "kzzk_cv.h"
#include "base_model.h"
#include "models/detection/yolov4.h"
#include "models/detection/yolov3.h"
#include "models/detection/vgg_ssd.h"
#include "models/detection/rt_detr.h"
#include "models/classification/resnet50.h"
#include "models/classification/mobilenetv1.h"
#include "models/classification/vgg16.h"
#include "models/segmentation/deeplabv3.h"

#include <memory>
#include <algorithm>
#include <cctype>
#include <iostream>
#include <sys/stat.h>

// ============================================================
// 静默模式开关：去掉所有 [INFO]/[DEBUG]/[WARN] 中间日志
// 需要调试时注释掉下面这行即可恢复完整日志输出
// ============================================================
#define KZZK_CV_SILENT_MODE

#ifdef KZZK_CV_SILENT_MODE
#  define KZZK_CERR   (void)0
#else
#  define KZZK_CERR   std::cerr
#endif

namespace kzzk_cv {

static std::string to_lower(const std::string& s) {
    std::string res = s;
    std::transform(res.begin(), res.end(), res.begin(),
                   [](unsigned char c) { return std::tolower(c); });
    return res;
}

static bool file_exists(const std::string& path) {
    struct stat st;
    return (stat(path.c_str(), &st) == 0) && S_ISREG(st.st_mode);
}

static std::string extract_model_name(const std::string& model_path) {
    std::string name = model_path;
    size_t last_slash = name.find_last_of("/\\");
    if (last_slash != std::string::npos) {
        name = name.substr(last_slash + 1);
    }
    size_t last_dot = name.find_last_of('.');
    if (last_dot != std::string::npos) {
        name = name.substr(0, last_dot);
    }
    return to_lower(name);
}

static std::unique_ptr<BaseModel> create_model(const std::string& model_name) {
    std::string lower = to_lower(model_name);

    if (lower.find("yolov4") != std::string::npos) {
        return std::make_unique<YOLOv4>();
    } else if (lower.find("yolov3") != std::string::npos) {
        return std::make_unique<YOLOv3>();
    } else if (lower.find("vgg_ssd") != std::string::npos || lower.find("vgg-ssd") != std::string::npos) {
        return std::make_unique<VGG_SSD>();
    } else if (lower.find("rtdetr") != std::string::npos ||
               lower.find("rt_detr") != std::string::npos ||
               lower.find("rt-detr") != std::string::npos) {
        return std::make_unique<RT_DETR>();
    } else if (lower.find("resnet50") != std::string::npos) {
        return std::make_unique<ResNet50>();
    } else if (lower.find("mobilenet") != std::string::npos) {
        return std::make_unique<MobileNetV1>();
    } else if (lower.find("vgg16") != std::string::npos) {
        return std::make_unique<VGG16>();
    } else if (lower.find("deeplabv3") != std::string::npos ||
               lower.find("deeplab") != std::string::npos) {
        return std::make_unique<DeepLabV3>();
    }

    return nullptr;
}

InferenceResult kzzk_cv(const std::string& modelfile, const std::string& imagefile) {
    InferenceResult result;
    result.infer_cost_ms = 0;

    // ---------- 第一步：文件存在性预检查 ----------
    if (!file_exists(modelfile)) {
        std::cerr << "[ERROR] 模型文件不存在: " << modelfile << std::endl;
        result.model_name = extract_model_name(modelfile);
        result.model_type = ModelType::DETECTION;
        return result;
    }
    if (!file_exists(imagefile)) {
        std::cerr << "[ERROR] 图像文件不存在: " << imagefile << std::endl;
        result.model_name = extract_model_name(modelfile);
        result.model_type = ModelType::DETECTION;
        return result;
    }

    // ---------- 第二步：按文件名匹配模型类 ----------
    std::string model_name = extract_model_name(modelfile);
    auto model = create_model(model_name);

    if (!model) {
        std::cerr << "[ERROR] 无法识别的模型类型: '" << model_name
                  << "' (从文件名 " << modelfile << " 解析)" << std::endl;
        std::cerr << "        支持的关键字: yolov4 / yolov3 / vgg_ssd / rt_detr / resnet50 / mobilenet / vgg16 / deeplabv3" << std::endl;
        result.model_name = model_name;
        result.model_type = ModelType::DETECTION;
        return result;
    }

    result.model_name = model->GetModelName();
    result.model_type = model->GetModelType();
    // ----- 静默模式：去掉 [INFO] 中间日志（取消注释调试时，需同步注释掉顶部 #define KZZK_CV_SILENT_MODE）-----
    // KZZK_CERR << "[INFO] 已识别模型: " << result.model_name << " (类型: "
    //           << static_cast<int>(result.model_type) << ")" << std::endl;
    // ------------------------------------------

    // ---------- 第三步：模型初始化 ----------
    if (!model->Initialize(modelfile, 0)) {
        std::cerr << "[ERROR] 模型初始化失败，请检查: " << std::endl;
        std::cerr << "        1) Ascend 驱动/环境是否就绪 (npu-smi info)" << std::endl;
        std::cerr << "        2) .om 模型文件是否损坏或与当前芯片架构匹配" << std::endl;
        std::cerr << "        3) 设备权限 (是否需要 root)" << std::endl;
        return result;
    }
    // ----- 静默模式：去掉 [INFO] 中间日志 -----
    // KZZK_CERR << "[INFO] 模型初始化成功" << std::endl;
    // ------------------------------------------

    // ---------- 第四步：执行推理 ----------
    result = model->Infer(imagefile);
    result.model_name = model->GetModelName();
    result.model_type = model->GetModelType();

    size_t total = result.detections.size() + result.classifications.size() + result.segmentation.class_info.size();
    // 真正失败的条件：耗时为 0 并且 **没有任何输出结果**（detection/classification/segmentation 全为空）
    // 如果有任一结果，哪怕耗时显示为 0ms 也只是因为 <1ms 被整数截断了，推理实际成功！
    if (result.infer_cost_ms == 0 && total == 0) {
        std::cerr << "[ERROR] 推理执行失败 (耗时为 0)，可能是图像预处理或模型执行阶段出错" << std::endl;
    } else {
        // ----- 静默模式：去掉 [INFO] 中间日志 -----
        // KZZK_CERR << "[INFO] 推理完成，耗时: " << result.infer_cost_ms << "ms" << std::endl;
        // KZZK_CERR << "[INFO] 输出结果条数: " << total << std::endl;
        // ------------------------------------------
    }

    model->Finalize();
    return result;
}

} // namespace kzzk_cv

File: cann/python/kzzk_cv.py
import sys
import os

path = os.path.dirname(os.path.abspath(__file__))
parent = os.path.dirname(path)
sys.path.insert(0, parent)

from python import kzzk_cv


def print_help():
    print("智能加速卡协同推理软件")
    print("Python 版命令行接口")
    print("Usage: kzzk_cv.py [OPTIONS]")
    print("Options:")
    print("  --modelfile MODEL      指定模型文件路径（.om 文件）")
    print("  --imagefile IMAGE      指定输入图像路径")
    print("  --help                 显示帮助信息")
    print("")
    print("支持的模型:")
    print("  YOLOv4, YOLOv3, VGG-SSD, RT-DETR  (目标检测)")
    print("  ResNet50, MobileNetV1, VGG16       (图像分类)")
    print("  DeepLabV3                          (图像分割)")
    print("")
    print("Examples:")
    print("  python kzzk_cv.py --modelfile /home/HwHiAiUser/cann/model/yolov4.om --imagefile /home/HwHiAiUser/cann/data/dog.jpg")
    print("  python kzzk_cv.py --modelfile /home/HwHiAiUser/cann/model/resnet50.om --imagefile /home/HwHiAiUser/cann/data/dog.jpg")


def main():
    args = sys.argv[1:]

    if not args or "--help" in args or "-h" in args:
        print_help()
        return

    modelfile = None
    imagefile = None

    i = 0
    while i < len(args):
        if args[i] == "--modelfile" and i + 1 < len(args):
            modelfile = args[i + 1]
            i += 2
        elif args[i] == "--imagefile" and i + 1 < len(args):
            imagefile = args[i + 1]
            i += 2
        else:
            i += 1

    if not modelfile or not imagefile:
        print("错误: 缺少必要参数", file=sys.stderr)
        print_help()
        sys.exit(1)

    try:
        result = kzzk_cv(modelfile, imagefile)
    except Exception as e:
        print("错误: %s" % str(e))
        sys.exit(1)

    print("Total inference cost: %dms" % result['infer_cost'])
    print("Inference Results: ")

    model_type = result['model_type']
    model_name = result['model_name']

    if model_type == 'detection':
        print("%s目标检测推理" % model_name)
        print("Total objects: %d" % len(result['results']))
        print("Format: [ID] (Class Name) [x1, y1, x2, y2] Confidence")
        for i, det in enumerate(result['results'], 1):
            bbox = det['bbox']
            print("[%d] %d (%s) [%d,%d,%d,%d] %d%%" % (
                i, det['class_id'], det['label'],
                bbox[0], bbox[1], bbox[2], bbox[3],
                int(round(det['confidence'] * 100))))

    elif model_type == 'classification':
        print("%s图像分类推理" % model_name)
        print("Top 5 classes:")
        print("Format: [ID Confidence]")
        for i, res in enumerate(result['results'], 1):
            print("Result %d: [%d %d%%]" % (
                i, res['class_id'], res['confidence']))

    elif model_type == 'segmentation':
        print("%s语义分割推理" % model_name)
        class_info = result['results']['class_info']
        print("Output classes count: %d" % len(class_info))
        print("Format: [ID] (Class Name) [Pixels] [Ratio]")
        for i, cls in enumerate(class_info, 1):
            print("Class %d: [%d] %d (%s) [%d] [%.2f%%]" % (
                i, cls['id'], cls['id'], cls['name'],
                cls['pixels'], cls['ratio']))

        seg_img_path = save_segmentation_image(result, imagefile)
        print("Segmentation image: %s" % seg_img_path)

    save_result_to_file(result, imagefile)


def save_segmentation_image(result, imagefile):
    from python.models.segmentation.deeplabv3 import DeepLabV3Segment
    out_dir = os.path.dirname(imagefile)
    if not out_dir:
        out_dir = "."
    base = os.path.splitext(os.path.basename(imagefile))[0]
    out_path = os.path.join(out_dir, base + "_seg.png")
    seg_map = result['results']['seg_map']
    DeepLabV3Segment.save_seg_image(imagefile, seg_map, out_path)
    return out_path


def save_result_to_file(result, imagefile):
    import datetime
    out_dir = os.path.dirname(imagefile)
    if not out_dir:
        out_dir = "."
    timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
    model_name = result['model_name']
    model_type = result['model_type']
    txt_path = os.path.join(out_dir, "%s_%s.txt" % (timestamp, model_name))

    with open(txt_path, 'w', encoding='utf-8') as f:
        f.write("Total inference cost: %dms\n" % result['infer_cost'])
        f.write("Inference Results:\n")

        if model_type == 'detection':
            f.write("%s目标检测推理\n" % model_name)
            f.write("Total objects: %d\n" % len(result['results']))
            f.write("Format: [ID] (Class Name) [x1, y1, x2, y2] Confidence\n")
            for i, det in enumerate(result['results'], 1):
                bbox = det['bbox']
                f.write("[%d] %d (%s) [%d,%d,%d,%d] %d%%\n" % (
                    i, det['class_id'], det['label'],
                    bbox[0], bbox[1], bbox[2], bbox[3],
                    int(round(det['confidence'] * 100))))

        elif model_type == 'classification':
            f.write("%s图像分类推理\n" % model_name)
            f.write("Top 5 classes:\n")
            f.write("Format: [ID Confidence]\n")
            for i, res in enumerate(result['results'], 1):
                f.write("Result %d: [%d %d%%]\n" % (
                    i, res['class_id'], res['confidence']))

        elif model_type == 'segmentation':
            seg_data = result['results']
            class_info = seg_data['class_info']
            f.write("%s语义分割推理\n" % model_name)
            f.write("Output classes count: %d\n" % len(class_info))
            f.write("Format: [ID] (Class Name) [Pixels] [Ratio]\n")
            for i, cls in enumerate(class_info, 1):
                f.write("Class %d: [%d] %d (%s) [%d] [%.2f%%]\n" % (
                    i, cls['id'], cls['id'], cls['name'],
                    cls['pixels'], cls['ratio']))

    print("推理结果已保存到: %s" % txt_path)


if __name__ == '__main__':
    main()

